sdks/go/pkg/beam/testing/teststream/teststream.go

```diff
@@ -0,0 +1,150 @@
+	c.watermark = timestamp
+	return nil
+}
+
+// AdvanceWatermarkToInfinity advances the watermark to the maximum timestamp.
+func (c *Config) AdvanceWatermarkToInfinity() error {
+	return c.AdvanceWatermark(mtime.MaxTimestamp.Milliseconds())
+}
+
+// AdvanceProcessingTime adds an event advancing the processing time by a given duration.
+// This advancement is applied to all of the PCollections output by the TestStream.
+func (c *Config) AdvanceProcessingTime(duration int64) {
+	processingAdvance := &pipepb.TestStreamPayload_Event_AdvanceProcessingTime{AdvanceDuration: duration}
+	processingEvent := &pipepb.TestStreamPayload_Event_ProcessingTimeEvent{ProcessingTimeEvent: processingAdvance}
+	c.events = append(c.events, &pipepb.TestStreamPayload_Event{Event: processingEvent})
+}
+
+// AdvanceProcessingTimeToInfinity moves the TestStream processing time to the largest possible
+// timestamp.
+func (c *Config) AdvanceProcessingTimeToInfinity() {
+	c.AdvanceProcessingTime(mtime.MaxTimestamp.Milliseconds())
+}
+
+// AddElements adds a number of elements to the Config object at the specified event timestamp.
+// The encoder will panic if there is a type mismatch between the provided coder and the
+// elements.
+func (c *Config) AddElements(timestamp int64, elements ...interface{}) error {
+	if c.elmCoder == nil {
+		elmType := reflect.TypeOf(elements[0])
+		var newCoder *coder.Coder
+		switch elmType.Kind() {
+		case reflect.Bool:
+			newCoder = coder.NewBool()
+		case reflect.String:
+			newCoder = coder.NewString()
+		case reflect.Float32, reflect.Float64:
+			newCoder = coder.NewDouble()
+		case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+			newCoder = coder.NewVarInt()
+		default:
+			return fmt.Errorf("unsupported element type %v", elmType.Kind())
+		}
+		c.elmCoder = newCoder
+	}
+	newElements := []*pipepb.TestStreamPayload_TimestampedElement{}
+	enc := beam.NewElementEncoder(c.elmCoder.T.Type())
+	for _, e := range elements {
+		var buf bytes.Buffer
+		if err := enc.Encode(e, &buf); err != nil {
+			return fmt.Errorf("encoding value %v failed, got %v", e, err)
+		}
+		newElements = append(newElements, &pipepb.TestStreamPayload_TimestampedElement{EncodedElement: buf.Bytes(), Timestamp: timestamp})
+	}
+	addElementsEvent := &pipepb.TestStreamPayload_Event_AddElements{Elements: newElements}
+	elementEvent := &pipepb.TestStreamPayload_Event_ElementEvent{ElementEvent: addElementsEvent}
+	c.events = append(c.events, &pipepb.TestStreamPayload_Event{Event: elementEvent})
+	return nil
+}
+
+// Create inserts a TestStream primitive into a pipeline, taking a scope and a Config object and
+// producing an array of output PCollections. The TestStream must be the first PTransform in the
+// pipeline.
+func Create(s beam.Scope, c Config) beam.PCollection {
+	pyld := protox.MustEncode(c.createPayload())
+	outputs := []beam.FullType{c.elmCoder.T}
+
+	outputMap := beam.External(s, urn, pyld, []beam.PCollection{}, outputs, false)
```

Comment:
Some cleanup from moving to `beam.External` from `beam.ExternalTagged`. `beam.External` already returns a `[]beam.PCollection`. Since the outputs are dictated by the the passed in slice of fulltypes, it should already only return a length 1 slice.

Reply:
Done.